lib/std/tar.zig

```diff
@@ -1,172 +1,1141 @@
+        .pax_buf = buf[block_len .. 2 * block_len],
+        .name_buf = buf[2 * block_len .. 3 * block_len],
+        .linkname_buf = buf[3 * block_len .. 4 * block_len],
+    };
+}
+
+pub fn HeaderIterator(comptime Reader: type) type {
+    return struct {
+        reader: Reader,
+        buf: Block = undefined,
+        pax_buf: Block = undefined,
+        pax_len: usize = 0,
+        name_buf: Block,
+        linkname_buf: Block,
+
+        const Self = @This();
+
+        /// iterates through the tar archive as if it is a series of
+        /// files. Internally, the tar format often uses fake "files" to add meta
+        /// data that describes the next file. These meta data "files" should not
+        /// normally be visible to the outside. As such, this iterates through
+        /// one or more "header files" until it finds a "normal file".
+        pub fn next(self: *Self) !?Header {
+            var gnu_long_name: []const u8 = "";
+            var gnu_long_link: []const u8 = "";
+            var format = fmt_ustar_pax_gnu;
+            var hdr = Header{};
+            var pax_hdr = Header{};
+
+            while (true) {
+                const v7 = try self.nextV7Header() orelse return null;
+                hdr = try self.header(v7) orelse return null;
+
+                format.setIntersection(hdr.fmt);
+                log.debug("hdr={}", .{hdr});
+                switch (hdr.type) {
+                    .extended_header, .global_extended_header => {
+                        format.setIntersection(fmt_pax);
+                        var paxiter = self.paxIterator();
+                        pax_hdr = .{};
+                        while (try paxiter.next()) |kv|
+                            try mergePax(kv, &pax_hdr);
+
+                        if (hdr.type == .global_extended_header) {
+                            var res = Header{
+                                .name = hdr.name,
+                                .type = hdr.type,
+                                .fmt = format,
+                            };
+                            res.merge(pax_hdr);
+                            return res;
+                        }
+                    },
+                    .gnu_long_name => {
+                        format.setIntersection(fmt_gnu);
+                        const amt = try self.reader.read(self.name_buf[0..self.buf.len]);
+                        if (amt != self.buf.len) return error.UnexpectedEndOfStream;
+                        gnu_long_name = mem.sliceTo(self.name_buf[0..self.buf.len], 0);
+                    },
+                    .gnu_long_link => {
+                        format.setIntersection(fmt_gnu);
+                        const amt = try self.reader.read(self.linkname_buf[0..self.buf.len]);
+                        if (amt != self.buf.len) return error.UnexpectedEndOfStream;
+                        gnu_long_link = mem.sliceTo(self.linkname_buf[0..self.buf.len], 0);
+                    },
```

Comment:
These are wrong, you need to check the length of the header for the size of the longname or linkname. They can easily be over 512 bytes. These headers use the "file content" of the header as value.

Reply:
Thanks. Dang, need a allocator now :shrug:

Reply:
went with an internal arena / page_allocator within `pipeToFileSystem()`. wasn't sure wether that function should accept one from the user.